OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/string_analysis/IntraproceduralStringAnalysis.scala

```diff
@@ -0,0 +1,369 @@
+ *
+ * @author Patrick Mell
+ */
+class IntraproceduralStringAnalysis(
+        val project: SomeProject
+) extends FPCFAnalysis {
+
+    /**
+     * This class is to be used to store state information that are required at a later point in
+     * time during the analysis, e.g., due to the fact that another analysis had to be triggered to
+     * have all required information ready for a final result.
+     */
+    private case class ComputationState(
+            // The lean path that was computed
+            computedLeanPath: Path,
+            // A mapping from DUVar elements to the corresponding indices of the FlatPathElements
+            var2IndexMapping: mutable.Map[V, Int],
+            // A mapping from values of FlatPathElements to StringConstancyInformation
+            fpe2sci: mutable.Map[Int, StringConstancyInformation],
+            // The three-address code of the method in which the entity under analysis resides
+            tac: TACode[TACMethodParameter, DUVar[ValueInformation]]
+    )
+
+    def analyze(data: P): ProperPropertyComputationResult = {
+        // sci stores the final StringConstancyInformation (if it can be determined now at all)
+        var sci = StringConstancyProperty.lb.stringConstancyInformation
+
+        val tacProvider = p.get(EagerDetachedTACAIKey)
+        val tac = tacProvider(data._2)
+
+        // Uncomment the following code to get the TAC from the property store
+        //        val tacaiEOptP = ps(data._2, TACAI.key)
+        //        var tac: TACode[TACMethodParameter, DUVar[ValueInformation]] = null
+        //        if (tacaiEOptP.hasUBP) {
+        //            if (tacaiEOptP.ub.tac.isEmpty) {
+        //                // No TAC available, e.g., because the method has no body
+        //                return Result(data, StringConstancyProperty.lb)
+        //            } else {
+        //                tac = tacaiEOptP.ub.tac.get
+        //            }
+        //        }
+        val cfg = tac.cfg
+        val stmts = tac.stmts
+
+        val uvar = data._1
+        val defSites = uvar.definedBy.toArray.sorted
+        // Function parameters are currently regarded as dynamic value; the following if finds read
+        // operations of strings (not String{Builder, Buffer}s, they will be handles further down
+        if (defSites.head < 0) {
+            return Result(data, StringConstancyProperty.lb)
+        }
+        val pathFinder: AbstractPathFinder = new WindowPathFinder(cfg)
+
+        // If not empty, this very routine can only produce an intermediate result
+        val dependees: mutable.Map[Entity, ListBuffer[EOptionP[Entity, Property]]] = mutable.Map()
+        // state will be set to a non-null value if this analysis needs to call other analyses /
+        // itself; only in the case it calls itself, will state be used, thus, it is valid to
+        // initialize it with null
+        var state: ComputationState = null
```

Comment:
The variable is only used&set in one branch. please just declare it there.

Reply:
As far as i see it, there is a usage of `state` in line 172, outside of the branch that sets the value. It seems to me there is a possible control-flow scenario where `state` is set (*or not set*) in the branch and then returned as part of the `continuation` in line 172.  Therefore i don't think i could simply move the declaration into the branch.

Reply:
State handling has been completely revamped since the `L0StringAnalysis` (formerly known as `IntraproceduralStringAnalysis` now requires proper dependency handling.